diff -uNr a/host_applications/linux/apps/hello_pi/hello_fft/makefile b/host_applications/linux/apps/hello_pi/hello_fft/makefile
--- a/host_applications/linux/apps/hello_pi/hello_fft/makefile	2014-11-29 02:45:48.000000000 +0800
+++ b/host_applications/linux/apps/hello_pi/hello_fft/makefile	2014-12-27 21:32:24.899116516 +0800
@@ -23,7 +23,7 @@
 F = -lrt -lm
 
 hello_fft.bin:	$(S) $(C1D) $(H1D)
-	gcc -o hello_fft.bin $(F) $(C1D)
+	gcc -o hello_fft.bin $(C1D) $(F)
 
 hello_fft_2d.bin:	$(S) hex/shader_trans.hex $(C2D) $(H2D)
 	gcc -o hello_fft_2d.bin $(F) $(C2D)
