diff --git a/xbmc/windowing/egl/EGLNativeType.h b/xbmc/windowing/egl/EGLNativeType.h
index aa5a1c5..27fc950 100644
--- a/xbmc/windowing/egl/EGLNativeType.h
+++ b/xbmc/windowing/egl/EGLNativeType.h
@@ -146,6 +146,12 @@ class CEGLNativeType
     framebuffer */
   virtual bool  ShowWindow(bool show) = 0;
 
+/*! \brief Waits for VSYNC, if possible
+
+    This call is there to wait until a VSYNC event happens, so that buffers
+    can be swapped without tearing, if the kernel driver implements it */
+  virtual void  WaitVSync() {};
+
 protected:
   XBNativeDisplayType  m_nativeDisplay;
   XBNativeWindowType   m_nativeWindow;
diff --git a/xbmc/windowing/egl/EGLWrapper.cpp b/xbmc/windowing/egl/EGLWrapper.cpp
index e19ce63..7c628a4 100644
--- a/xbmc/windowing/egl/EGLWrapper.cpp
+++ b/xbmc/windowing/egl/EGLWrapper.cpp
@@ -374,6 +374,7 @@ bool CEGLWrapper::SetVSync(EGLDisplay display, bool enable)
   // might fail so let caller decide if this is an error.
   status = eglSwapInterval(display, enable ? 1 : 0);
   CheckError();
+  m_vsync = (enable && m_result == EGL_SUCCESS);
   return status;
 }
 
@@ -381,6 +382,8 @@ void CEGLWrapper::SwapBuffers(EGLDisplay display, EGLSurface surface)
 {
   if ((display == EGL_NO_DISPLAY) || (surface == EGL_NO_SURFACE))
     return;
+  if (m_vsync)
+    m_nativeTypes->WaitVSync();
   eglSwapBuffers(display, surface);
 }
 
diff --git a/xbmc/windowing/egl/EGLWrapper.h b/xbmc/windowing/egl/EGLWrapper.h
index cd4bbb5..7d7513f 100644
--- a/xbmc/windowing/egl/EGLWrapper.h
+++ b/xbmc/windowing/egl/EGLWrapper.h
@@ -72,4 +72,5 @@ class CEGLWrapper
 
   CEGLNativeType *m_nativeTypes;
   EGLint         m_result;
+  bool           m_vsync;
 };
