--- vdr/dvbplayer.c	2015/02/01 10:45:41	3.3
+++ vdr/dvbplayer.c	2015/02/02 09:39:11
@@ -377,6 +377,8 @@
   if (index) {
      int Index = ptsIndex.FindIndex(DeviceGetSTC());
      if (Index >= 0) {
+        if (Setup.SkipEdited && marks.First() && abs(Index - marks.First()->Position()) <= int(round(RESUMEBACKUP * framesPerSecond)))
+           Index = 0; // when stopping within RESUMEBACKUP of the first mark the recording shall still be considered unviewed
         Index -= int(round(RESUMEBACKUP * framesPerSecond));
         if (Index > 0)
            Index = index->GetNextIFrame(Index, false);
