diff -uNr vdr-graphtftng-0.4.10+r6459127.orig/graphtft-fe/Makefile vdr-graphtftng-0.4.10+r6459127/graphtft-fe/Makefile
--- vdr-graphtftng-0.4.10+r6459127.orig/graphtft-fe/Makefile	2014-08-13 13:29:03.308125552 +0800
+++ vdr-graphtftng-0.4.10+r6459127/graphtft-fe/Makefile	2014-08-13 13:31:55.792980862 +0800
@@ -4,7 +4,7 @@
 CXXFLAGS  += -Wno-deprecated-declarations
 LFLAGS    = -Wl,--no-undefined
 LIBS      = -lpthread -ljpeg -lX11
-LIBS      += $(shell imlib2-config --libs)
+LIBS      += $(shell pkg-config imlib2 --libs)
 AR        = ar
 
 TARGET = graphtft-fe
