diff -uNr gcc-linaro-4.9-2015.01.orig/gcc/Makefile.in gcc-linaro-4.9-2015.01/gcc/Makefile.in
--- gcc-linaro-4.9-2015.01.orig/gcc/Makefile.in	2015-02-02 11:07:23.959775510 +0800
+++ gcc-linaro-4.9-2015.01/gcc/Makefile.in	2015-02-02 11:15:53.806303702 +0800
@@ -334,7 +334,7 @@
 ISLINC = @ISLINC@
 
 # How to find CLOOG
-CLOOGLIBS = @CLOOGLIBS@
+CLOOGLIBS = @CLOOGLIBS@ -lcloog-isl -lisl
 CLOOGINC = @CLOOGINC@
 
 # Set to 'yes' if the LTO front end is enabled.
@@ -964,7 +964,7 @@
 # programs built during a bootstrap.
 # autoconf inserts -DCROSS_DIRECTORY_STRUCTURE if we are building a
 # cross compiler which does not use the native headers and libraries.
-INTERNAL_CFLAGS = -DIN_GCC $(PICFLAG) @CROSS@
+INTERNAL_CFLAGS = -DIN_GCC -DCLOOG_INT_GMP $(PICFLAG) @CROSS@
 
 # This is the variable actually used when we compile. If you change this,
 # you probably want to update BUILD_CFLAGS in configure.ac
