diff --git a/arch/arm/plat-meson/power.c b/arch/arm/plat-meson/power.c
index e73df57..185fda88 100644
--- a/arch/arm/plat-meson/power.c
+++ b/arch/arm/plat-meson/power.c
@@ -41,7 +41,9 @@ void meson_common_restart(char mode,const char *cmd)
 {
     u32 reboot_reason = MESON_NORMAL_BOOT;
     if (cmd) {
-        if (strcmp(cmd, "charging_reboot") == 0)
+        if (strcmp(cmd, "poweroff") == 0)
+            reboot_reason = MESON_UBOOT_SUSPEND;
+        else if (strcmp(cmd, "charging_reboot") == 0)
             reboot_reason = MESON_CHARGING_REBOOT;
         else if (strcmp(cmd, "recovery") == 0 || strcmp(cmd, "factory_reset") == 0)
             reboot_reason = MESON_FACTORY_RESET_REBOOT;
@@ -71,8 +73,7 @@ void meson_power_off_prepare(void)
 
 void meson_power_off(void)
 {
-	printk("meson power off \n");
-	meson_common_restart('h',"charging_reboot");
+        meson_common_restart('h', "poweroff");
 }
 
 void meson_power_idle(void)
diff --git a/include/uapi/linux/reboot.h b/include/uapi/linux/reboot.h
index 973bf32..7137376 100755
--- a/include/uapi/linux/reboot.h
+++ b/include/uapi/linux/reboot.h
@@ -60,6 +60,7 @@
 #define	MESON_LOCK_REBOOT				0x08080808
 #define	MESON_USB_BURNER_REBOOT			0x09090909
 #define	MESON_REBOOT_CLEAR				0xdeaddead
+#define	MESON_UBOOT_SUSPEND                    0x0b0b0b0b
 #endif /* CONFIG_PLAT_MESON */
 
 #endif /* _UAPI_LINUX_REBOOT_H */
