diff --git a/drivers/amlogic/wifi/broadcm_40181/dhd_config.c b/drivers/amlogic/wifi/broadcm_40181/dhd_config.c
index 0da1622..388bc0e 100755
--- a/drivers/amlogic/wifi/broadcm_40181/dhd_config.c
+++ b/drivers/amlogic/wifi/broadcm_40181/dhd_config.c
@@ -108,6 +108,13 @@ const static char *bcm4339a0ag_fw_name[] = {
 	"fw_bcm4339a0_ag_p2p.bin",
 	"fw_bcm4339a0_ag_mfg.bin"
 };
+
+const static char *nv_name[] = {
+	"ap6210_nvram.txt",
+	"ap6330_nvram.txt",
+	"ap6181_nvram.txt",
+	"ap6335e_nvram.txt",
+};
 
 #define htod32(i) i
 #define htod16(i) i
@@ -322,6 +329,67 @@ dhd_conf_set_nv_name_by_mac(dhd_pub_t *dhd, bcmsdh_info_t *sdh, char *nv_path)
 	}
 }
 
+void
+dhd_conf_set_nv_name_by_chip(dhd_pub_t *dhd, char *dst, char *src)
+{
+	int fw_type, ag_type;
+	static uint chip, chiprev, first=1;
+	int i;
+
+	if (first) {
+		chip = dhd_bus_chip_id(dhd);
+		chiprev = dhd_bus_chiprev_id(dhd);
+		first = 0;
+	}
+
+	if (src[0] == '\0') {
+#ifdef CONFIG_BCMDHD_NVRAM_PATH
+		bcm_strncpy_s(src, sizeof(nvram_path), CONFIG_BCMDHD_NVRAM_PATH, MOD_PARAM_PATHLEN-1);
+		if (src[0] == '\0')
+#endif
+		{
+			printf("src nvram path is null\n");
+			return;
+		}
+	}
+
+		strcpy(dst, src);
+	#ifndef FW_PATH_AUTO_SELECT
+		return;
+	#endif
+
+	/* find out the last '/' */
+	i = strlen(dst);
+	while (i>0){
+		if (dst[i] == '/') break;
+		i--;
+	}
+	switch (chip) {
+		case BCM4330_CHIP_ID:
+					strcpy(&dst[i+1], nv_name[1]);
+						break;
+		case BCM43362_CHIP_ID:
+					strcpy(&dst[i+1], nv_name[0]);
+						break;	
+		case BCM43340_CHIP_ID: //bcm43341b0ag BCM43340B0
+					strcpy(&dst[i+1], nv_name[0]);
+						break;
+//		case BCM43341_CHIP_ID: //bcm43341b0ag BCM43341B0 
+//					strcpy(&dst[i+1], nv_name[0]);
+//						break;
+//		case BCM4324_CHIP_ID: //bcm43241b4ag BCM43241B4
+//					strcpy(&dst[i+1], nv_name[0]);
+//						break;
+		case BCM4335_CHIP_ID: //bcm4339a0ag BCM4335A0
+					strcpy(&dst[i+1], nv_name[3]);
+						break;
+		case BCM4339_CHIP_ID: //bcm4339a0ag
+					strcpy(&dst[i+1], nv_name[3]);
+						break;
+	}
+	printf("%s: nvram_path=%s\n", __FUNCTION__, dst);
+}
+
 void
 dhd_conf_set_fw_name_by_chip(dhd_pub_t *dhd, char *dst, char *src)
 {
diff --git a/drivers/amlogic/wifi/broadcm_40181/dhd_config.h b/drivers/amlogic/wifi/broadcm_40181/dhd_config.h
index 003f997..f603fb9 100755
--- a/drivers/amlogic/wifi/broadcm_40181/dhd_config.h
+++ b/drivers/amlogic/wifi/broadcm_40181/dhd_config.h
@@ -15,6 +15,7 @@
 
 #define FW_PATH_AUTO_SELECT 1
 extern char firmware_path[MOD_PARAM_PATHLEN];
+extern char nvram_path[MOD_PARAM_PATHLEN];
 extern int disable_proptx;
 extern uint dhd_doflow;
 #ifdef POWER_OFF_IN_SUSPEND
@@ -111,6 +112,7 @@ typedef struct dhd_conf {
 extern void *bcmsdh_get_drvdata(void);
 void dhd_conf_set_fw_name_by_mac(dhd_pub_t *dhd, bcmsdh_info_t *sdh, char *fw_path);
 void dhd_conf_set_nv_name_by_mac(dhd_pub_t *dhd, bcmsdh_info_t *sdh, char *nv_path);
+void dhd_conf_set_nv_name_by_chip(dhd_pub_t *dhd, char *dst, char *src);
 void dhd_conf_set_fw_name_by_chip(dhd_pub_t *dhd, char *dst, char *src);
 #if defined(HW_OOB)
 void dhd_conf_set_hw_oob_intr(bcmsdh_info_t *sdh, uint chip);
diff --git a/drivers/amlogic/wifi/broadcm_40181/dhd_linux.c b/drivers/amlogic/wifi/broadcm_40181/dhd_linux.c
index 39c829e..c56f514 100755
--- a/drivers/amlogic/wifi/broadcm_40181/dhd_linux.c
+++ b/drivers/amlogic/wifi/broadcm_40181/dhd_linux.c
@@ -2936,6 +2936,7 @@ dhd_open(struct net_device *net)
 		if (firmware_path[strlen(firmware_path)-1] == '\n')
 			firmware_path[strlen(firmware_path)-1] = '\0';
 		dhd_conf_set_fw_name_by_chip(&dhd->pub, fw_path, firmware_path);
+		dhd_conf_set_nv_name_by_chip(&dhd->pub, nv_path, nvram_path);
 	}
 
 
@@ -3258,8 +3259,10 @@ dhd_attach(osl_t *osh, struct dhd_bus *bus, uint bus_hdrlen)
 	/* Link to bus module */
 	dhd->pub.bus = bus;
 	dhd->pub.hdrlen = bus_hdrlen;
-	if (strlen(firmware_path) != 0)
+	if (strlen(firmware_path) != 0){
 		dhd_conf_set_fw_name_by_chip(&dhd->pub, fw_path, firmware_path);
+		dhd_conf_set_nv_name_by_chip(&dhd->pub, nv_path, nvram_path);
+	}
 
 	/* Set network interface name if it was provided as module parameter */
 	if (iface_name[0]) {
diff --git a/drivers/amlogic/wifi/broadcm_40181/dhd_sdio.c b/drivers/amlogic/wifi/broadcm_40181/dhd_sdio.c
index 8ee717d..a008541 100755
--- a/drivers/amlogic/wifi/broadcm_40181/dhd_sdio.c
+++ b/drivers/amlogic/wifi/broadcm_40181/dhd_sdio.c
@@ -8466,6 +8466,7 @@ dhd_bus_devreset(dhd_pub_t *dhdp, uint8 flag)
 			                        bus->cl_devid)) {
 				/* Attempt to download binary to the dongle */
 				dhd_conf_set_fw_name_by_chip(dhdp, fw_path, firmware_path); // terence
+				dhd_conf_set_nv_name_by_chip(dhdp, nv_path, nvram_path);
 				if (dhdsdio_probe_init(bus, bus->dhd->osh, bus->sdh) &&
 					dhdsdio_download_firmware(bus, bus->dhd->osh, bus->sdh)) {
 
@@ -8519,6 +8520,7 @@ dhd_bus_devreset(dhd_pub_t *dhdp, uint8 flag)
 			dhd_conf_set_hw_oob_intr(bus->sdh, bus->sih->chip); // terence 20120615: fix for OOB initial issue
 #endif
 			dhd_conf_set_fw_name_by_chip(dhdp, fw_path, firmware_path);
+			dhd_conf_set_nv_name_by_chip(dhdp, nv_path, nvram_path);
 			if ((bcmerror = dhd_bus_start(dhdp)) != 0)
 				DHD_ERROR(("%s: dhd_bus_start fail with %d\n",
 					__FUNCTION__, bcmerror));
