diff -uNr systemd-218.orig/src/core/manager.c systemd-218/src/core/manager.c
--- systemd-218.orig/src/core/manager.c	2015-02-16 11:02:20.744093160 +0800
+++ systemd-218/src/core/manager.c	2015-02-16 11:02:56.512091664 +0800
@@ -1878,11 +1878,6 @@
                          * set to yes.
                          *
                          * See http://bugs.debian.org/624599 */
-                        if (strcmp(j->unit->id, "sysinit.target") == 0 && unit->default_dependencies) {
-                                log_debug("sysinit.target is queued, ignoring %s request for unit %s", job_type_to_string(type), unit->id);
-                                sd_bus_error_setf(e, SD_BUS_ERROR_INVALID_ARGS, "sysinit.target is queued, ignoring %s request for unit %s", job_type_to_string(type), unit->id);
-                                return -EINVAL;
-                        }
                 }
                 }
         }
diff -uNr systemd-218.orig/src/tmpfiles/tmpfiles.c systemd-218/src/tmpfiles/tmpfiles.c
--- systemd-218.orig/src/tmpfiles/tmpfiles.c	2015-02-16 11:02:20.724093161 +0800
+++ systemd-218/src/tmpfiles/tmpfiles.c	2015-02-16 11:03:14.048090930 +0800
@@ -39,7 +39,7 @@
 #include <glob.h>
 #include <fnmatch.h>
 #include <sys/capability.h>
-#include <sys/xattr.h>
+#include <attr/xattr.h>
 
 #include "log.h"
 #include "util.h"
